InfrastructureManager/appscale/infrastructure/__init__.py

```diff
@@ -0,0 +1,522 @@
+      raise CustomHTTPError(HTTPCodes.UNAUTHORIZED, message='Invalid secret')
+
+    parameters = json_decode(self.request.body)
+    logger.info('Operation id received: {}'.format(parameters))
+    if not parameters or 'operation_id' not in parameters:
+      raise CustomHTTPError(HTTPCodes.BAD_REQUEST, message='operation_id is a'
+                                                           'required parameter')
+
+    operation_id = parameters['operation_id']
+    if operation_ids.has_key(operation_id):
+      self.write(json_encode(operation_ids.get(operation_id)))
+    else:
+      logger.error('Operation id not found')
+      raise CustomHTTPError(HTTPCodes.NOT_FOUND, message='Operation id not '
+                                                         'found')
+  @gen.coroutine
+  def post(self):
+    """
+    Spawn new instances using the provided parameters. The request must also
+    contain a 'num_vms' parameter which indicates the number of virtual
+    machines that should be spawned.
+
+    Args:
+      AppScale-Secret: Required in header. Authentication to deployment.
+      args: The request body, a dictionary of values needed to start
+        instances in the specified infrastructure. Infrastructure specific
+        checks are made in  'get_agent', but the following parameters will be
+        verified here:
+          num_vms: Required in body. Number of VMs to start.
+          infrastructure: Required in body, infrastructure to construct an agent
+            for.
+    Returns:
+      If the secret is valid and all the required parameters are available in
+      the input parameter map, this method will return a dictionary containing
+      an 'operation_id' key.
+
+    Raises:
+      CustomHTTPError if the necessary parameters are not filled.
+    """
+    if 'AppScale-Secret' not in self.request.headers \
+        or self.request.headers['AppScale-Secret'] != options.secret:
+      raise CustomHTTPError(HTTPCodes.UNAUTHORIZED, message='Invalid secret')
+
+    args = json_decode(self.request.body)
+
+    logger.info('Received a request to run instances.')
+    for arg in [PARAM_INFRASTRUCTURE, PARAM_NUM_VMS]:
+      if arg not in args:
+        raise CustomHTTPError(HTTPCodes.BAD_REQUEST,
+              message='{} is a required parameter'.format(arg))
+
+    num_vms = int(args[PARAM_NUM_VMS])
+    if num_vms <= 0:
+      logger.warn('Invalid VM count: {0}'.format(num_vms))
+      raise CustomHTTPError(HTTPCodes.BAD_REQUEST,
+                            message='Invalid VM count: {0}'.format(num_vms))
+
+    try:
+      agent, run_params = get_agent(self.agent_factory,
+                                    BaseAgent.OPERATION_RUN, args)
+    except AgentConfigurationException as exception:
+      logger.exception("Error creating agent!")
+      raise CustomHTTPError(HTTPCodes.INTERNAL_ERROR,
+        message='Invalid agent configuration: {0}'.format(exception))
+
+    operation_id = str(uuid.uuid4())
```

Comment:
It could be nice to generate the operation id early and then include it in all logging related to that operation.

Reply:
I looked at this a bit more and realized that everything up until the line where the operation_id is currently assigned is configuration checking/error handling. It's extremely likely that if one request fails due to any of the above config checks then all will, so I'm not certain if an operation id would be very helpful. Additionally we have no need to save information about an operation id if it fails above since it does not actually make a call that we need to check on.